src/transformers/generation/utils.py

```diff
@@ -2001,11 +2001,26 @@ def _prepare_cache_for_generation(
         # Use DynamicCache() instance by default. This will avoid back and forth from legacy format that
         # keeps copying the cache thus using much more memory
         else:
-            model_kwargs[cache_name] = (
-                DynamicCache(**dynamic_cache_kwargs)
-                if not requires_cross_attention_cache
-                else EncoderDecoderCache(DynamicCache(**dynamic_cache_kwargs), DynamicCache(**dynamic_cache_kwargs))
-            )
+            if not requires_cross_attention_cache:
+                model_kwargs[cache_name] = DynamicCache(**dynamic_cache_kwargs)
+            else:
+                # For encoder-decoder models, we need to use separate configs for encoder and decoder
+                decoder_cache_kwargs = dynamic_cache_kwargs.copy()
+                encoder_cache_kwargs = dynamic_cache_kwargs.copy()
+
+                if "config" in dynamic_cache_kwargs:
+                    # Access the encoder/decoder sub-configs directly for models like T5Gemma
+                    if hasattr(self.config, "decoder") and hasattr(self.config, "encoder"):
+                        decoder_cache_kwargs["config"] = self.config.decoder
+                        encoder_cache_kwargs["config"] = self.config.encoder
+                    else:
```

Comment:
This shouldn't be needed -- if `self.config.get_text_config` is not working properly, then it means it is missing logic :)

(I think `get_text_config()` is missing `"encoder"` in the `encoder_possible_text_config_names` list)